### PR TITLE
Add WebAuthn as a second factor

### PR DESCRIPTION
**Pull request by klaustopher**
- [x] Add [webauthn-ruby](https://github.com/cedarcode/webauthn-ruby) gem
- [x] Add logic
  - [x] Add new columns into `two_factor_authentication_devices` column to track ID, public key and sign in count
  - [x] Database model and registering in the `TokenStrategyManager`
  - [x] Registering a device
  - [x] Authentication with a device
- [x] Frontend for adding a new device 
- [x] Frontend for authenticating in login process
- [x] Implement controller logic for the `modules/two_factor_authentication/app/controllers/two_factor_authentication/users/two_factor_devices_controller.rb`
  - [x] Forbid switching to other auth device types when creating for others (currently changing type to totp in the url allows setting up a totp)
- [x] Implement controller logic for the `modules/two_factor_authentication/app/controllers/two_factor_authentication/forced_registration/two_factor_devices_controller.rb`
- [x] We currently display the raw error message to the client. At least for the most common error `AbortError` (when the user clicked cancel in the dialog that pops up) we should display an i18n'd message.
- [ ] Add tests
---

Implements https://community.openproject.org/work_packages/48743/activity

### CODE REVIEW

**Review by oliverguenther — CHANGES_REQUESTED** [review]
Looking mostly good, I just noticed this:

When I don't have any 2FA set up, creating a new TOTP devices and confirming it will result in the device being confirmed _and_ the default, resulting in 2FA enabled.

I just created a webauthn device, but it's not marked as default and 2FA is not active until I manually clicked on "make default". I assume this is due to you skipping the confirmation phase implicitly 
Bildschirmfoto 2024-03-07 um 08 46 16

File: modules/two_factor_authentication/lib/open_project/two_factor_authentication/token_strategy_manager.rb [hunk 1]
```diff
@@ -114,7 +114,7 @@ def enforced_by_configuration?
       def merge_with_settings!(config)
         config['active_strategies'] ||= []
         # Always enable totp if nothing is enabled
-        config['active_strategies'] << :totp if add_default_strategy?(config)
+        config['active_strategies'] += %i[totp] if add_default_strategy?(config)
```
Comment (oliverguenther) [inline modules/two_factor_authentication/lib/open_project/two_factor_authentication/token_strategy_manager.rb]:
```suggestion
        config['active_strategies'] += %i[totp webauthn] if add_default_strategy?(config)
```

Webauthn should be a default strategy so people can use it directly

**Comment by klaustopher** [comment]
> I just created a webauthn device, but it's not marked as default and 2FA is not active until I manually clicked on "make default". I assume this is due to you skipping the confirmation phase implicitly

Oh, I fixed this only in one place (forced registration). Will add in the other one as well

File: modules/two_factor_authentication/lib/open_project/two_factor_authentication/token_strategy_manager.rb (hunk 1 above)
Comment (klaustopher) [reply modules/two_factor_authentication/lib/open_project/two_factor_authentication/token_strategy_manager.rb]: Did not want to put any new defaults on people, but makes sense. reverted bc5665a57d20b41f00f3861ad1fb5c2d71cfa491

**Review by oliverguenther — APPROVED** [review]

**Pull request merged**